src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+ngx_http_lua_semaphore_handler(ngx_event_t *ev)
+{
+    ngx_http_lua_semaphore_t          *sem;
+    ngx_http_request_t                *r;
+    ngx_http_lua_ctx_t                *ctx;
+    ngx_http_lua_co_ctx_t             *wait_co_ctx;
+    ngx_connection_t                  *c;
+    ngx_queue_t                       *q;
+
+    sem = ev->data;
+
+    while (!ngx_queue_empty(&sem->wait_queue) && sem->resource_count > 0) {
+
+        q = ngx_queue_head(&sem->wait_queue);
+        ngx_queue_remove(q);
+
+        sem->wait_count--;
+
+        wait_co_ctx = ngx_queue_data(q, ngx_http_lua_co_ctx_t, sem_wait_queue);
+        wait_co_ctx->cleanup = NULL;
+
+        if (wait_co_ctx->sleep.timer_set) {
+            ngx_del_timer(&wait_co_ctx->sleep);
+        }
+
+        r = ngx_http_lua_get_req(wait_co_ctx->co);
+        c = r->connection;
+
+        ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+        if (ctx == NULL) {
+            continue;
+        }
+
+        sem->resource_count--;
+
+        ctx->cur_co_ctx = wait_co_ctx;
+
+        wait_co_ctx->sem_resume_status = SEMAPHORE_WAIT_SUCC;
+
+        if (ctx->entered_content_phase) {
+            (void) ngx_http_lua_semaphore_resume(r);
+
+        } else {
+            ctx->resume_handler = ngx_http_lua_semaphore_resume;
+            ngx_http_core_run_phases(r);
+        }
+
+        ngx_http_run_posted_requests(c);
+    }
+
+    /* after dealing with semaphore post event */
+    sem->in_post_event = 0;
+}
+
+
+static void
+ngx_http_lua_semaphore_timeout_handler(ngx_event_t *ev)
+{
+    ngx_http_lua_co_ctx_t       *wait_co_ctx;
+    ngx_http_request_t          *r;
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_connection_t            *c;
+    ngx_http_lua_semaphore_t    *sem;
+
+    wait_co_ctx = ev->data;
+    wait_co_ctx->cleanup = NULL;
+
+    dd("ngx_http_lua_semaphore_timeout_handler timeout coctx:%p", wait_co_ctx);
+
+    sem = wait_co_ctx->data;
+
+    ngx_queue_remove(&wait_co_ctx->sem_wait_queue);
+    sem->wait_count--;
+
+    r = ngx_http_lua_get_req(wait_co_ctx->co);
+    c = r->connection;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return;
+    }
+
+    ctx->cur_co_ctx = wait_co_ctx;
+
+    wait_co_ctx->sem_resume_status = SEMAPHORE_WAIT_TIMEOUT;
+
+    if (ctx->entered_content_phase) {
+        (void) ngx_http_lua_semaphore_resume(r);
+
+    } else {
+         ctx->resume_handler = ngx_http_lua_semaphore_resume;
+         ngx_http_core_run_phases(r);
+    }
+
+    ngx_http_run_posted_requests(c);
+}
+
+
+void
+ngx_http_lua_ffi_semaphore_gc(ngx_http_lua_semaphore_t *sem)
+{
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                   "ngx_http_lua_ffi_semaphore_gc %p", sem);
+
+    if (sem == NULL) {
+        return;
+    }
+
+    if (!ngx_queue_empty(&sem->wait_queue)) {
+        ngx_log_error(NGX_LOG_CRIT, ngx_cycle->log, 0,
+                       "ngx_http_lua_ffi_semaphore_gc wait queue is"
```

Comment:
Style: arguments on the continued lines should align up with the first argument of the function call.